overlord/hookstate/hookmgr.go

```diff
@@ -153,13 +150,19 @@ func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 
 	handler := handlers[0]
 
+	context, err := m.activeHandlers.addHandler(handler)
```

Comment:
There's a bit of terminology confusion here which would be good to avoid. What we have here is not the context. It's the context ID, and the actual context is in line 142 above:
```
handlers := m.repository.generateHandlers(&Context{task: task, setup: setup})
```
That value being created inline and passed as a parameter is the actual context, and we should probably get it with NewContext(task, setup) so it has a chance to generate its own ID internally. We can then:
```
m.context[context.ID()] = context
```
So, map of context, a context ID, and the context.